File: database/src/main/java/cc/imorning/database/utils/MessageDatabaseHelper.java
package cc.imorning.database.utils;

import android.content.Context;
import android.util.Log;

import java.util.HashMap;
import java.util.Map;

import cc.imorning.database.BuildConfig;
import cc.imorning.database.db.MessageDB;

public class MessageDatabaseHelper {

    private static final String TAG = "MessageDatabaseHelper";

    private static volatile MessageDatabaseHelper messageDatabaseHelper;
    private static Map<String, MessageDB> messageDBMap;

    private MessageDatabaseHelper() {
        messageDBMap = new HashMap<>();
    }

    public static MessageDatabaseHelper getInstance() {
        if (messageDatabaseHelper == null) {
            synchronized (MessageDatabaseHelper.class) {
                if (messageDatabaseHelper == null) {
                    messageDatabaseHelper = new MessageDatabaseHelper();
                }
            }
        }
        return messageDatabaseHelper;
    }

    public void addNewInstance(Context context, String user, String me) {

        if (messageDBMap.containsKey(me)) {
            return;
        }
        // MessageDB messageDB = Room.databaseBuilder(
        //         context,
        //         MessageDB.class,
        //         DatabaseHelper.INSTANCE.getMessageDatabase(context, user, me)
        // ).build();
        MessageDB messageDB = MessageDB.Companion.getInstance(context, user, me);
        messageDBMap.put(user, messageDB);
    }

    public MessageDB getMessageDatabaseDao(Context context, String user, String me) {
        if (messageDBMap.get(user) == null) {
            addNewInstance(context, user, me);
        }
        if (BuildConfig.DEBUG) {
            Log.d(TAG, "getMessageDatabaseDao: " + messageDBMap.get(user));
        }
        return messageDBMap.get(user);
    }

}

File: database/src/main/java/cc/imorning/database/utils/MessageDatabaseHelper.kt
package cc.imorning.database.utils

import android.content.Context
import cc.imorning.database.db.MessageDB
import cc.imorning.database.db.MessageDB.Companion.getInstance

class MessageDatabaseHelper private constructor() {
    init {
        messageDBMap = HashMap()
    }

    fun initNewMessageDBInstance(context: Context?, user: String, me: String): MessageDB? {
        if (messageDBMap.containsKey(me)) {
            return messageDBMap[me]
        }
        val messageDB = getInstance(context!!, user, me)
        messageDBMap[user] = messageDB
        return messageDB
    }

    /**
     * Get the message database,create it if the object is not cached
     */
    fun getMessageDB(context: Context?, user: String, me: String): MessageDB? {
        var messageDB = messageDBMap[user]
        if (messageDB == null) {
            messageDB = initNewMessageDBInstance(context, user, me)
        }
        return messageDB
    }

    companion object {
        private const val TAG = "MessageDatabaseHelper"

        @Volatile
        private var messageDatabaseHelper: MessageDatabaseHelper? = null

        // use HashMap to cache MessageDB instance
        private lateinit var messageDBMap: MutableMap<String, MessageDB>

        val instance: MessageDatabaseHelper?
            get() {
                if (messageDatabaseHelper == null) {
                    synchronized(MessageDatabaseHelper::class.java) {
                        if (messageDatabaseHelper == null) {
                            messageDatabaseHelper = MessageDatabaseHelper()
                        }
                    }
                }
                return messageDatabaseHelper
            }
    }
}